File: paleo_workbench/ui/pages/preview_cache.py
from __future__ import annotations

from collections import OrderedDict
from pathlib import Path

from paleo_workbench.project.models import ExportArtifact, ResourceItem
from paleo_workbench.ui.pages.preview_provider import PreviewResult

DEFAULT_PREVIEW_CACHE_BYTES = 128 * 1024 * 1024


def preview_result_weight(value: PreviewResult) -> int:
    """Return the payload bytes retained by a cached preview result."""
    if value.estimated_bytes > 0:
        return value.estimated_bytes
    return (
        len(value.text.encode("utf-8"))
        + len(value.image_bytes)
        + len(value.pdf_bytes)
    )


def safe_file_stat(path: Path) -> tuple[int, int] | None:
    """Return ``(size, mtime_ns)`` for cache keys, or None if the path is unreadable."""
    try:
        st = path.stat()
        return (st.st_size, getattr(st, "st_mtime_ns", int(st.st_mtime * 1e9)))
    except OSError:
        return None


# Backward-compatible alias for internal call sites / older imports.
_safe_stat = safe_file_stat


def make_preview_cache_key(
    asset: ResourceItem | ExportArtifact,
    settings_fingerprint: str | None = None,
) -> tuple:
    """Stable key for preview LRU entries.

    Includes type/format so rescan reclassification (same path/stat) is a miss.
    Filesystem (size, mtime_ns) invalidates on rewrite; checksum when present.
    """
    if settings_fingerprint is None:
        from paleo_workbench.ui.pages.preview_settings import PreviewSettings

        settings_fingerprint = PreviewSettings.defaults().fingerprint()

    if isinstance(asset, ExportArtifact):
        path = Path(asset.output_path)
        return (
            "artifact",
            asset.id,
            asset.output_path,
            asset.format,
            "",
            safe_file_stat(path),
            settings_fingerprint,
        )
    path = Path(asset.path)
    return (
        "resource",
        asset.id,
        asset.path,
        asset.type,
        asset.format,
        asset.checksum or "",
        safe_file_stat(path),
        settings_fingerprint,
    )


class PreviewCache:
    def __init__(
        self,
        max_size: int = 32,
        max_bytes: int = DEFAULT_PREVIEW_CACHE_BYTES,
    ):
        for name, value in (("max_size", max_size), ("max_bytes", max_bytes)):
            if not isinstance(value, int) or isinstance(value, bool):
                raise TypeError(f"{name} must be an integer")
            if value <= 0:
                raise ValueError(f"{name} must be positive")
        self.max_size = max_size
        self.max_bytes = max_bytes
        self.current_bytes = 0
        self._data: OrderedDict[tuple, tuple[PreviewResult, int]] = OrderedDict()

    def get(self, key: tuple) -> PreviewResult | None:
        if key not in self._data:
            return None
        self._data.move_to_end(key)
        return self._data[key][0]

    def put(self, key: tuple, value: PreviewResult) -> None:
        if key in self._data:
            _old_value, old_weight = self._data.pop(key)
            self.current_bytes -= old_weight

        weight = preview_result_weight(value)
        if weight > self.max_bytes:
            return

        self._data[key] = (value, weight)
        self.current_bytes += weight
        while len(self._data) > self.max_size or self.current_bytes > self.max_bytes:
            _old_key, (_old_value, old_weight) = self._data.popitem(last=False)
            self.current_bytes -= old_weight

    def clear(self) -> None:
        self._data.clear()
        self.current_bytes = 0

File: libs/ui_data_core/src/preview_cache.cpp
// preview_cache.py port — see preview_cache.hpp.

#include <chrono>
#include "pwb/ui_data_core/preview_cache.hpp"

#include "pwb/ui_data_core/json_util.hpp"

#include <stdexcept>
#include <sys/stat.h>

namespace pwb::ui_data_core {

long long preview_result_weight(const PreviewResult& value) {
    if (value.estimated_bytes > 0) {
        return value.estimated_bytes;
    }
    return static_cast<long long>(value.text.size()) +
           static_cast<long long>(value.image_bytes.size()) +
           static_cast<long long>(value.pdf_bytes.size());
}

std::optional<std::pair<long long, long long>> safe_file_stat(
    const std::filesystem::path& path) {
#if defined(_WIN32)
    // MSVC stat() takes narrow paths only; fs probes give the same facts.
    std::error_code ec{};
    if (!std::filesystem::is_regular_file(path, ec) || ec) {
        return std::nullopt;
    }
    const long long size =
        static_cast<long long>(std::filesystem::file_size(path, ec));
    if (ec) return std::nullopt;
    const auto written = std::filesystem::last_write_time(path, ec);
    if (ec) return std::nullopt;
    const long long mtime_ns = std::chrono::duration_cast<
        std::chrono::nanoseconds>(written.time_since_epoch())
        .count();
    return std::make_pair(size, mtime_ns);
#else
    struct stat st {};
    if (::stat(path.c_str(), &st) != 0) {
        return std::nullopt;
    }
    long long mtime_ns;
#ifdef st_mtime  // glibc macro → st_mtim.tv_sec
    mtime_ns = static_cast<long long>(st.st_mtim.tv_sec) * 1000000000LL +
               static_cast<long long>(st.st_mtim.tv_nsec);
#else
    mtime_ns = static_cast<long long>(st.st_mtime) * 1000000000LL;
#endif
    return std::make_pair(static_cast<long long>(st.st_size), mtime_ns);
#endif
}

namespace {

// Join tuple fields into one canonical key string. json.dumps(value,
// ensure_ascii=True) gives an unambiguous encoding per field; "\x1f"
// separates fields (it cannot appear inside a dumps-encoded string).
std::string key_join(std::initializer_list<domain::Json> fields) {
    std::string out;
    bool first = true;
    for (const auto& field : fields) {
        if (!first) {
            out += '\x1f';
        }
        first = false;
        out += field.dump(-1, ' ', true);
    }
    return out;
}

domain::Json stat_json(
    const std::optional<std::pair<long long, long long>>& stat) {
    if (!stat.has_value()) {
        return domain::Json(nullptr);
    }
    return domain::Json::array({stat->first, stat->second});
}

// metadata.get("coordinate_units") or metadata.get("units") or "" — the
// raw value (any JSON type) goes into the key tuple, not str(value).
domain::Json metadata_units(const domain::Json& metadata) {
    if (!metadata.is_object()) {
        return "";
    }
    if (const auto it = metadata.find("coordinate_units");
        it != metadata.end() && json_truthy(*it)) {
        return *it;
    }
    if (const auto it = metadata.find("units");
        it != metadata.end() && json_truthy(*it)) {
        return *it;
    }
    return "";
}

// metadata.get("comparison_crs") or "" — raw value or "".
domain::Json metadata_comparison(const domain::Json& metadata) {
    if (!metadata.is_object()) {
        return "";
    }
    if (const auto it = metadata.find("comparison_crs");
        it != metadata.end() && json_truthy(*it)) {
        return *it;
    }
    return "";
}

std::string fingerprint_or_default(
    const std::optional<std::string>& fingerprint) {
    if (fingerprint.has_value()) {
        return *fingerprint;
    }
    return PreviewSettings::defaults().fingerprint();
}

}  // namespace

std::string make_preview_cache_key(
    const ExportArtifact& asset,
    std::optional<std::string> settings_fingerprint,
    std::optional<std::string> /*comparison_crs*/) {
    const auto stat = safe_file_stat(std::filesystem::path(asset.output_path));
    return key_join({
        "artifact",
        asset.id,
        asset.output_path,
        asset.format,
        "",
        stat_json(stat),
        fingerprint_or_default(settings_fingerprint),
    });
}

std::string make_preview_cache_key(
    const ResourceItem& asset,
    std::optional<std::string> settings_fingerprint,
    std::optional<std::string> comparison_crs) {
    const auto stat = safe_file_stat(std::filesystem::path(asset.path));
    const domain::Json& metadata =
        asset.parsed_summary.is_object() ? asset.parsed_summary
                                         : domain::Json::object();
    domain::Json comparison;
    if (comparison_crs.has_value()) {
        comparison = *comparison_crs;
    } else {
        comparison = metadata_comparison(metadata);
    }
    return key_join({
        "resource",
        asset.id,
        asset.path,
        asset.type,
        asset.format,
        asset.checksum.value_or(""),
        stat_json(stat),
        asset.crs.value_or(""),
        metadata_units(metadata),
        comparison,
        fingerprint_or_default(settings_fingerprint),
    });
}

std::string make_preview_cache_key_for_view(
    const AssetView& view,
    std::optional<std::string> settings_fingerprint,
    std::optional<std::string> comparison_crs) {
    const auto stat = safe_file_stat(std::filesystem::path(view.path));
    const domain::Json& metadata =
        view.parsed_summary.is_object() ? view.parsed_summary
                                        : domain::Json::object();
    domain::Json comparison;
    if (comparison_crs.has_value()) {
        comparison = *comparison_crs;
    } else {
        comparison = metadata_comparison(metadata);
    }
    return key_join({
        "resource",
        view.id,
        view.path,
        view.type,
        view.format,
        view.checksum.value_or(""),
        stat_json(stat),
        view.crs.value_or(""),
        metadata_units(metadata),
        comparison,
        fingerprint_or_default(settings_fingerprint),
    });
}

// ---------------------------------------------------------------------------
// PreviewCache
// ---------------------------------------------------------------------------

PreviewCache::PreviewCache(long long max_size, long long max_bytes)
    : max_size_(max_size), max_bytes_(max_bytes) {
    if (max_size <= 0) {
        throw std::invalid_argument("max_size must be positive");
    }
    if (max_bytes <= 0) {
        throw std::invalid_argument("max_bytes must be positive");
    }
}

const PreviewResult* PreviewCache::get(const std::string& key) {
    const auto it = map_.find(key);
    if (it == map_.end()) {
        return nullptr;
    }
    lru_.splice(lru_.end(), lru_, it->second);  // move_to_end
    return &it->second->second.first;
}

void PreviewCache::put(const std::string& key, const PreviewResult& value) {
    put(key, PreviewResult(value));
}

void PreviewCache::put(const std::string& key, PreviewResult&& value) {
    if (const auto it = map_.find(key); it != map_.end()) {
        current_bytes_ -= it->second->second.second;
        lru_.erase(it->second);
        map_.erase(it);
    }
    const long long weight = preview_result_weight(value);
    if (weight > max_bytes_) {
        return;
    }
    lru_.emplace_back(key, std::make_pair(std::move(value), weight));
    map_[key] = std::prev(lru_.end());
    current_bytes_ += weight;
    while (static_cast<long long>(map_.size()) > max_size_ ||
           current_bytes_ > max_bytes_) {
        auto& oldest = lru_.front();
        current_bytes_ -= oldest.second.second;
        map_.erase(oldest.first);
        lru_.pop_front();
    }
}

void PreviewCache::clear() {
    lru_.clear();
    map_.clear();
    current_bytes_ = 0;
}

}  // namespace pwb::ui_data_core
